### PR TITLE
Optimize graph partition

### PR DESCRIPTION
**Pull request by snnn**
Before this change,
If there are 3 providers, GPU, CPU, mkldnn
Then for each node, KernelRegistry::TryFindKernel will be called 3x3 = 9 times, instead of 3.
N providers need O(N^2) searches. It should be O(N).

Perf result:

**Baseline**:

Benchmark                  |Time          | CPU |Iterations
------------------------|------------|-------|------------
BM_SearchKernelRegistry_SingleNodeModel            |245 ns        |244 ns  |172212113
BM_SearchKernelRegistry_RealModel_tiny_yolo       |9551 ns       |9540 ns    |4412595
BM_SearchKernelRegistry_RealModel_inception_v4       |149148 ns       |148933 ns    |281947
BM_PartitionModel_tiny_yolo       |1971998 ns    |1961591 ns       |1435
BM_PartitionModel_inception_v4    |4453991 ns    |4443086 ns        |626

New code:

Benchmark                                                                     |Time                |Time Diff   |  CPU |Iterations
----------------------------------------------------------- |----------------|-------|-------|-------------------
BM_SearchKernelRegistry_SingleNodeModel                |291 ns             | +19%|291 ns  |143936682
BM_SearchKernelRegistry_RealModel_tiny_yolo            |11401 ns         |+19% |11389 ns    |3681155
BM_SearchKernelRegistry_RealModel_inception_v4      |138977 ns       |-7%    | 138789 ns    |302019
BM_PartitionModel_tiny_yolo                                        |1686156 ns     |-16%  | 1676349 ns       |1673
BM_PartitionModel_inception_v4                                  |1683208 ns     |-62%  |1676687 ns       |1656

For each kernel search call, it's about 20 percent slower.
However, overall, the graph partition process is significantly faster. And the new implementation works much better on complex models.

### CODE REVIEW

File: include/onnxruntime/core/framework/tensor.h [hunk 1]
```diff
@@ -164,8 +164,15 @@ class Tensor final {
     shape_ = new_shape;
   }
 
+  //return zero if integer overflow
   size_t Size() const noexcept {
```
Comment (skottmckay) [inline include/onnxruntime/core/framework/tensor.h]: Would it be better to throw an exception if there's an error calculating the size? I assume you don't expect that to happen, and returning 0 could accidentally be interpreted as an empty tensor, delaying when the error due to this breaks something (making it much harder to debug).

File: include/onnxruntime/core/session/onnxruntime_cxx_api.h [hunk 2]
```diff
@@ -7,16 +7,23 @@
...
+#define ORT_THROW_ON_ERROR(expr)                                                   \
+  do {                                                                             \
+    OrtStatus* onnx_status = (expr);                                               \
+    if (onnx_status != nullptr) {                                                  \
+      std::string ort_error_message = OrtGetErrorMessage(onnx_status);             \
+      OrtErrorCode error_code = OrtGetErrorCode(onnx_status);                      \
+      OrtReleaseStatus(onnx_status);                                               \
+      switch (error_code) {                                                        \
+        case ORT_NOT_IMPLEMENTED:                                                  \
+          throw ::onnxruntime::NotImplementedException(ort_error_message);         \
```
Comment (skottmckay) [inline include/onnxruntime/core/session/onnxruntime_cxx_api.h]: >:: [](start = 16, length = 2)

nit: Using the '::' prefix shouldn't be necessary. We should not have any 'onnxruntime' namespaces nested within another, so just 'onnxruntime::' should be unambiguous.

File: onnxruntime/core/framework/allocation_planner.cc [hunk 3]
```diff
@@ -195,13 +195,11 @@ class PlannerImpl {
   // Find if there exists some input tensor that we can use in-place for output_arg
   bool FindReusableInput(const onnxruntime::Node& node, int output_arg_num, MLValueIndex* reusable_input) {
     auto p_output_arg = node.OutputDefs()[output_arg_num];
-    auto p_opkernel_def = utils::GetKernelDef(kernel_registry_, node);
+    const KernelCreateInfo* ci;
+    Status st = kernel_registry_.SearchKernelRegistry(node, &ci);
+    if (!st.IsOK() || ci == nullptr || ci->kernel_def == nullptr) return false;
```
Comment (skottmckay) [inline onnxruntime/core/framework/allocation_planner.cc]: nit: I know the google c++ style allows this, but putting the return statement on the same line as the 'if' means you have to reformat the code to put a breakpoint on the 'return' statement if you wanted to debug when the return was used.

File: onnxruntime/core/framework/allocation_planner.cc [hunk 4]
```diff
@@ -363,15 +361,19 @@ class PlannerImpl {
         if (node_input->Exists())
           UseCount(node_input->Name())++;
       }
-
       // Identify where each output of this node should be allocated.
       // This is determined by the opkernel bound to the node.
+      //TODO: does 'pnode' equal to 'node' ? Can we remove the following line?
       auto node = graph_viewer_.GetNode(step.node_index);
       ORT_ENFORCE(nullptr != node);
```
Comment (skottmckay) [inline onnxruntime/core/framework/allocation_planner.cc]: I would expect it to be the same as I don't believe step.node_index changes here.

File: onnxruntime/core/framework/allocation_planner.cc [hunk 5]
```diff
@@ -419,11 +421,12 @@ class PlannerImpl {
     return Status::OK();
   }
 
-  void GeneratePlanForWeights() {
+  //TODO: Don't generate plan for CPU tensors?
+  Status GeneratePlanForWeights() {
```
Comment (skottmckay) [inline onnxruntime/core/framework/allocation_planner.cc]: It's usually helpful to include the 'why' and not just 'what' so the reason for making such a change is captured. e.g. Don't generate plan for CPU tensors because ...

File: onnxruntime/core/framework/graph_partitioner.cc [hunk 6]
```diff
@@ -80,12 +80,10 @@ Status GraphPartitioner::Partition(Graph& graph, bool export_dll, FuncManager& f
...
   std::shared_ptr<KernelRegistry> fused_kernel_registry = std::make_shared<KernelRegistry>();
   // Partitioning <graph> based on provider preference and their capabilities.
-  auto kernel_registries = kernel_registry_mgr_.GetAllKernelRegistries();
-
   std::vector<std::vector<std::unique_ptr<ComputeCapability>>> capabilities_of_all_providers;
   GraphViewer graph_viewer(graph);
   for (auto& provider : providers_) {
-    capabilities_of_all_providers.push_back(provider->GetCapability(graph_viewer, kernel_registries));
+    capabilities_of_all_providers.push_back(provider->GetCapability(graph_viewer, kernel_registry_mgr_.GetKernelRegistriesByProviderType(provider->Type())));
   }
```
Comment (skottmckay) [inline onnxruntime/core/framework/graph_partitioner.cc]: nit: line too long. 120 is the max. [https://marketplace.visualstudio.com/items?itemName=PaulHarrington.EditorGuidelines](https://marketplace.visualstudio.com/items?itemName=PaulHarrington.EditorGuidelines) is very helpful as you can set a guideline at 120 and see if something is too long.

It's not enforced by clang-format so that developers can choose where to put line breaks for better readability (vs. clang-format always trying to fit as much as possible on a line).

File: onnxruntime/core/framework/kernel_registry.cc [hunk 7]
```diff
@@ -113,25 +113,6 @@ class TypeBindingResolver {
...
-// type-constraints of the given kernel. This serves two purposes: first, to
-// select the right kernel implementation based on the types of the arguments
-// when we have multiple kernels, e.g., Clip<float> and Clip<int>; second, to
-// accommodate (and check) mapping of ONNX (specification) type to the onnxruntime
-// implementation type (e.g., if we want to implement ONNX's float16 as a regular
-// float in onnxruntime). (The second, however, requires a globally uniform mapping.)
-//
-// Note that this is not intended for type-checking the node against the ONNX
-// type specification of the corresponding op, which is done before this check.
 bool KernelRegistry::VerifyKernelDef(const onnxruntime::Node& node,
```
Comment (skottmckay) [inline onnxruntime/core/framework/kernel_registry.cc]: Why delete the comment?

File: onnxruntime/core/framework/kernel_registry_manager.cc [hunk 8]
```diff
@@ -14,63 +14,105 @@ Status KernelRegistryManager::CreateKernel(const onnxruntime::Node& node,
...
+    status = p->CreateKernel(node,
+                             execution_provider,
+                             session_state.GetInitializedTensors(),
+                             session_state.GetMLValueNameIdxMap(),
+                             session_state.GetFuncMgr(),
+                             op_kernel);
     if (status.IsOK()) {
       return status;
     }
   }
```
Comment (skottmckay) [inline onnxruntime/core/framework/kernel_registry_manager.cc]: A comment explaining why this second phase of kernel creation is needed would be helpful. Also, we call CreateKernel under the lock on line 25, but outside of the lock on line 44. Why the differenece?

File: onnxruntime/core/framework/kernel_registry_manager.cc [hunk 9]
```diff
@@ -14,63 +14,105 @@ Status KernelRegistryManager::CreateKernel(const onnxruntime::Node& node,
...
-    status = registry->CreateKernel(node,
-                                    execution_provider,
-                                    session_state.GetInitializedTensors(),
-                                    session_state.GetMLValueNameIdxMap(),
-                                    session_state.GetFuncMgr(),
-                                    op_kernel);
+  {
+    std::lock_guard<OrtMutex> lock(lock_);
+    for (auto& registry : kernel_registries_) {
+      status = registry->CreateKernel(node,
```
Comment (skottmckay) [inline onnxruntime/core/framework/kernel_registry_manager.cc]: >CreateKernel [](start = 25, length = 12)

Side question: would TryCreateKernel be a better name?

File: onnxruntime/core/framework/kernel_registry_manager.cc [hunk 10]
```diff
@@ -14,63 +14,105 @@ Status KernelRegistryManager::CreateKernel(const onnxruntime::Node& node,
...
 }
 
-void KernelRegistryManager::RegisterKernels(const ExecutionProviders& execution_providers,
-                                            KernelRegistryPriority priority) {
-  for (auto& provider : execution_providers)
-    RegisterKernelRegistry(provider->GetKernelRegistry(), priority);
+void KernelRegistryManager::RegisterKernels(const ExecutionProviders& execution_providers) {
+  for (auto& provider : execution_providers) {
+    std::lock_guard<OrtMutex> lock(lock_);
+    auto iter = provider_type_to_registry_.find(provider->Type());
```
Comment (skottmckay) [inline onnxruntime/core/framework/kernel_registry_manager.cc]: Is there a reason to keep acquiring/freeing the lock within the loop instead of outside it?

Comment (snnn) [reply onnxruntime/core/framework/kernel_registry_manager.cc]: Very tiny difference. 
Because the code
```c++
for (auto& provider : execution_providers)
```
will try to invoke some functions in ExecutionProviders class. If any of these functions tried to get the lock in KernelRegistryManager again,  it will be a deadlock.
Usually, I would prefer to release a lock as soon as possible to avoid deadlock, but in this case, as you pointed, there is no such need.

File: include/onnxruntime/core/framework/tensor.h (hunk 1 above)
Comment (snnn) [reply include/onnxruntime/core/framework/tensor.h]: The function is marked as 'noexcept'. If I change it to throw exception, I need to change a lot different places.

File: include/onnxruntime/core/session/onnxruntime_cxx_api.h (hunk 2 above)
Comment (snnn) [reply include/onnxruntime/core/session/onnxruntime_cxx_api.h]: This is a public API file.  We won't do that, but the other people would.

File: onnxruntime/core/framework/kernel_registry_manager.h [hunk 11]
```diff
@@ -17,46 +18,52 @@ class KernelRegistry;
...
 // The 1st and 2nd ones are shared across sessions.
+
+//This class is thread safe
 class KernelRegistryManager {
  public:
   KernelRegistryManager() = default;
 
-  void RegisterKernels(const ExecutionProviders& execution_providers,
-                       KernelRegistryPriority priority = KernelRegistryPriority::LowPriority);
+  void RegisterKernels(const ExecutionProviders& execution_providers);
```
Comment (skottmckay) [inline onnxruntime/core/framework/kernel_registry_manager.h]: Does this create a more confusing setup? You now have things registered by RegisterKernelRegistry and this method in separate members. Based on the current implementation, values in kernel_registries_ will by higher priority than anything registered here. 

Why do the two sets of KernelRegistry instances need to be separate?

Comment (snnn) [reply onnxruntime/core/framework/kernel_registry_manager.h]: They need to have different priorities. Currently, all the custom registries have higher priority than any provider. So, it would happen that:
1. user registered a kernel for GEMM op, but the kernel is only run on CPU
2. However, he prefer to use GPU, he registered CUDA provider before the CPU provider
3.  As the priority is defined as this, the cpu one will be used.

Comment (snnn) [reply onnxruntime/core/framework/kernel_registry_manager.h]: Inside the implementation, we may not need two collections, just one is enough.

Comment (snnn) [reply onnxruntime/core/framework/kernel_registry_manager.h]: Let me think about it.

**Comment by linkerzhang** [comment]
quick question: can we not regress the perf on small model (first 3 rows)? :)

**Comment by snnn** [comment]
> quick question: can we not regress the perf on small model (first 3 rows)? :)

The top 3 rows are for KernelRegistryManager::SearchKernelRegistry function. Whether it's small model or large model, it's always slower.

The other rows are for GraphPartioner ,which will call KernelRegistryManager::SearchKernelRegistry function many many times. Even KernelRegistryManager::SearchKernelRegistry is slower,  GraphPartioner is much faster.

File: onnxruntime/core/framework/kernel_registry.cc (hunk 7 above)
Comment (snnn) [reply onnxruntime/core/framework/kernel_registry.cc]: Moved it to the header file

File: onnxruntime/core/framework/kernel_registry_manager.cc (hunk 9 above)
Comment (snnn) [reply onnxruntime/core/framework/kernel_registry_manager.cc]: Why?

File: onnxruntime/core/framework/kernel_registry_manager.h (hunk 11 above)
Comment (snnn) [reply onnxruntime/core/framework/kernel_registry_manager.h]: I didn't change the behavior. It still works exactly as before. 
The kernels either are from a provider, or from a custom registry. 
custom registries always have higher priority than the builtin ones.

**Comment by snnn** [comment]
Hi @skottmckay, would you like to review it again?

File: include/onnxruntime/core/framework/tensor.h (hunk 1 above)
Comment (skottmckay) [reply include/onnxruntime/core/framework/tensor.h]: Why? I thought noexcept was mainly an optimization opportunity and didn't require changes at the calling site. Suddenly returning zero seems like it would actually need all the calling sites to be aware of that possibility and handle it, so I would have thought throwing would have less impact on other code. 

---
In reply to: [255722659](https://github.com/Microsoft/onnxruntime/pull/465#discussion_r255722659) [](ancestors = 255722659)

File: onnxruntime/core/framework/kernel_registry_manager.cc (hunk 9 above)
Comment (skottmckay) [reply onnxruntime/core/framework/kernel_registry_manager.cc]: Because it only attempts to create the kernel if found by TryFindKernel. If I see CreateKernel as the name, I would not think failure was expected. But in this case we're attempting to create a kernel if found, and expect failures. 

---
In reply to: [256098369](https://github.com/Microsoft/onnxruntime/pull/465#discussion_r256098369) [](ancestors = 256098369)

File: onnxruntime/core/framework/session_state_initializer.cc [hunk 12]
```diff
@@ -366,38 +366,27 @@ common::Status SaveInitializedTensors(const onnxruntime::Graph& graph,
...
-
+  common::Status status = custom_registry_manager.CreateKernel(node,
+                                                               *exec_provider,
+                                                               session_state,
+                                                               op_kernel);
   if (!status.IsOK()) {
-    LOGS(logger, ERROR) << "Kernel creation failed for node: "
-                        << node.Name() << " with error: " << status.ErrorMessage();
+    return common::Status(status.Category(), status.Code(),
+                          ::onnxruntime::MakeString("Kernel creation failed for node: ", node.Name(), " with error: ",
```
Comment (skottmckay) [inline onnxruntime/core/framework/session_state_initializer.cc]: >:: [](start = 26, length = 2)

Do we need to use this qualifier? We should never have an 'onnxruntime' namespace nested within another namespace if things are setup correctly, so the '::' should not be required.

File: onnxruntime/core/framework/kernel_registry_manager.h [hunk 13]
```diff
@@ -17,46 +18,56 @@ class KernelRegistry;
...
       result.push_back(registry.get());
     }
+    auto iter = provider_type_to_registry_.find(type);
+    if (iter != provider_type_to_registry_.end())
+      result.push_back(iter->second.get());
     return result;
   }
-
- private:
   ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(KernelRegistryManager);
```
Comment (skottmckay) [inline onnxruntime/core/framework/kernel_registry_manager.h]: nit: prefer whitespace separating unrelated code. Also does ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE need to move to the 'public' section?

File: onnxruntime/core/framework/kernel_registry_manager.h (hunk 11 above)
Comment (skottmckay) [reply onnxruntime/core/framework/kernel_registry_manager.h]: Would it be clearer if kernel_registries_ was called custom_kernel_registries_ and there was a comment before the member declaration explaining why that's separate from the values in provider_type_to_registry_ and how the two different collections are used? This is mainly to help developers understand the structure of the class. 

---
In reply to: [256193230](https://github.com/Microsoft/onnxruntime/pull/465#discussion_r256193230) [](ancestors = 256193230)

File: onnxruntime/core/framework/kernel_registry_manager.h [hunk 14]
```diff
@@ -17,46 +18,56 @@ class KernelRegistry;
...
                       /*out*/ std::unique_ptr<OpKernel>& op_kernel) const;
 
+  //This function assumes the node is already assigned to an execution provider
+  //Don't call this function before graph partition is done
   Status SearchKernelRegistry(const onnxruntime::Node& node,
                               /*out*/ const KernelCreateInfo** kernel_create_info) const;
 
+  bool HasImplementationOf(const Node& node, const std::string& provider_type) const;
+
   // Get all kernel registries. There are no nullptr entries.
```
Comment (skottmckay) [inline onnxruntime/core/framework/kernel_registry_manager.h]: >Get all kernel registries [](start = 5, length = 25)

This comment doesn't really describe the code does it? Seems like this returns all custom registries, and the registry for the given provider type.

File: onnxruntime/core/framework/kernel_registry_manager.cc (hunk 9 above)
Comment (snnn) [reply onnxruntime/core/framework/kernel_registry_manager.cc]: As you said, usually it shouldn't fail.

Comment (skottmckay) [reply onnxruntime/core/framework/kernel_registry_manager.cc]: I didn't intend to. :) What I meant was we only attempt to create it if TryFindKernel is successful. Calling this TryCreateKernel would be more consistent with that. If found, we don't expect it to fail, but lookups in custom registries would (I assume) fail more often than not. 

---
In reply to: [258171590](https://github.com/Microsoft/onnxruntime/pull/465#discussion_r258171590) [](ancestors = 258171590)

Comment (snnn) [reply onnxruntime/core/framework/kernel_registry_manager.cc]: This function is called after graph partition, after the kernel search. So, we expect we can find it.

Comment (snnn) [reply onnxruntime/core/framework/kernel_registry_manager.cc]: KernelRegistryManager::CreateKernel calls KernelRegistry::CreateKernel,which calls KernelRegistry::TryFindKernel.

KernelRegistry::TryFindKernel is also called in the graph partition phrase.

File: onnxruntime/core/framework/kernel_registry_manager.h (hunk 13 above)
Comment (snnn) [reply onnxruntime/core/framework/kernel_registry_manager.h]: "Deleted member function should be public"
See: 

1. https://abseil.io/tips/143
2. https://google.github.io/styleguide/cppguide.html#Copyable_Movable_Types
"deleting the appropriate operations in the **public** section of the declaration"

**Comment by snnn** [comment]
Hi @skottmckay

Any comments ?

File: onnxruntime/core/framework/kernel_registry_manager.h (hunk 13 above)
Comment (skottmckay) [reply onnxruntime/core/framework/kernel_registry_manager.h]: Cool. Thanks for the pointers - learnt something. I was just following the previous usages of doing that in the 'private' section. Good to know why doing it in 'public' is better.
